fs: restore javascript realpath implementation

bzoz opened the pull request:
<!--
Thank you for your pull request. Please review below requirements.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test nosign` (Windows) passes
- [x] commit message follows commit guidelines

##### Affected core subsystem(s)
fs

##### Description of change
Libuv `realpath` implementation brought a lot of problems described here https://github.com/nodejs/node/issues/7726. This restores old JS implementation of `realpath` and `realpathSync`, adding new API, with `options` in place of `cache`.

This requires #7846 to land, as it reverts changes to `fs` on which old JS implementation depended. This PR is about 3rd and 4th commit which bring back old `realpath` implementation with support for the new API.

First two commits are from https://github.com/nodejs/node/pull/7846, reverting some changes to `fs` that landed after JS `realpath` implementation was dropped.

Third commit reverts https://github.com/nodejs/node/commit/b488b19eaf2b2e7a3ca5eccd2445e245847a5f76 bringing back JS implementation of `realpath` and `realpathSync`.

Fourth commit updates old implementation. It removes the cache and implements new `realpath` API (support for `encoding` parameter).

This cannot land before #7846. In case #7846 does not make it, this can be easily updated.

addaleax commented:
A couple of suggestions:
- I don’t think the native binding to the libuv realpath needs to be removed. It can be, but it might be nice to retain it for comparing results with the JS implementation later.
- You could add the benchmark/test commit from @trevnorris’ #7548 here.
- I’ll try to benchmark it, but removing the cache altogether sounds like something that won’t work out. I realize that it conflicts with the `encoding:` option, but that seems to be easily worked around by never adding an `encoding:` key to the cache? It also won’t be a likely problem in practice, because only `path.resolve()`d keys are added to the cache.
- The encoding option should probably be used for all `fs.*` calls that are made from inside the realpath implementations?

orangemocha commented:
CI run: https://ci.nodejs.org/job/node-test-pull-request/3441/

orangemocha commented:
At the last meeting, the CTC has expressed consensus with the approach in this PR: reverting, leaving the cache out, but keeping the encoding option.

ChALkeR commented:
To me it looks like this reverts too many tests. Could we keep the ones that pass?

bzoz commented:
@addaleax :
- I think that since we are not using native realpath anymore there is no need to keep bindings around. 
- Adding benchmarks from #7548 is a good idea, I'll do that.
- Inside we are using `path.resolve`, which expects utf-8 strings. So we stick to them when calling other `fs` functions, and only apply encoding option when returning result.

@ChALkeR: Those test are reverted by #7846 on which this PR depends.

addaleax commented:
> I think that since we are not using native realpath anymore there is no need to keep bindings around. 

@bzoz Your call.There seems to be agreement that the long-term plan is to use a fixed libuv implementation, though, and I think keeping the binding around might be useful for e.g. comparative benchmarks.

>  we are using `path.resolve`, which expects utf-8 strings

`path.resolve` expects strings, which means that it it isn’t concerned with encodings at all. But you’re right, this is not as easy as I though it would be, and it’s probably best to just stick with UTF-8 for all intermediate results (and accepting that that means that symlinks encountered along the way must be valid UTF-8).

jasnell commented:
In general this LGTM with a green CI. I definitely want to make sure this gets additional eyes on it to review, however. @nodejs/ctc

StefanScherer commented:
@bzoz I've just compiled your PR on Windows and tried the `node.exe` inside a Windows Container. Installed Node.js 6.3.1 from MSI inside the container and just replaced `node.exe`. 

Running `npm init`, `npm install express` ... works fine again. 🐳

saghul commented:
Why does this PR also revert "fs: validate args of truncate functions in js" and "fs: make callback mandatory to all async functions" ? How are those related to the realpath problems? It makes is a lot harder to review this PR.

saghul commented:
Those seem to be handled by https://github.com/nodejs/node/pull/7846, btw.

addaleax commented:
@saghul This PR would otherwise create merge conflicts, that’s all; the intent here is to wait until #7846 is landed, if I understood @bzoz correctly.

addaleax commented on test/parallel/test-fs-realpath-on-substed-drive.js:
```diff
@@ -0,0 +1,60 @@
+  assert(!err);
+  assert.equal(typeof result, 'string');
+  assert.equal(result, filename);
+});
```
You don’t need the `asyncCompleted` checks; wrapping the callback as `common.mustCall(function(err, result) { … })` handles that automatically for you.

addaleax commented on test/parallel/test-fs-realpath-on-substed-drive.js:
```diff
@@ -0,0 +1,60 @@
+const driveLetters = 'ABCDEFGHIJKLMNOPQRSTUWXYZ';
+let driveLetter;
+for (var i = 0; i < driveLetters.length; ++i) {
+  driveLetter = `${driveLetters[i]}:`;
```
I’d maybe just call this variable `drive`… `driveLetter` sounds so much like a single character ;)

addaleax commented on test/parallel/test-fs-realpath-on-substed-drive.js:
```diff
@@ -0,0 +1,60 @@
+
+result = fs.realpathSync(filename);
+assert.equal(typeof result, 'string');
+assert.equal(result, filename);
```
nit: use `assert.strictEqual` wherever it makes sense

addaleax commented on test/parallel/test-fs-realpath.js:
```diff
@@ -197,7 +197,7 @@ function test_cyclic_link_protection(callback) {
     unlink.push(t[0]);
   });
-  assert.throws(function() { fs.realpathSync(entry); }, /ELOOP/);
+  assert.throws(function() { fs.realpathSync(entry); });
```
Just curious, why this change? If it’s because of something Windows-specific, maybe amending the regex here makes sense?

bzoz replied:
I just reverted all changes to test-fs-realpath from @trevnorris commit. It was specific for C realpath implementation.

addaleax replied:
@bzoz Ah, I see now – are you sure these changes were specific to the C realpath implementation? I know I requested this particular change here (adding `/ELOOP/`) because this test had also passed when an error other than `ELOOP` was thrown (thus masquerading a real bug); all implemetations should throw `ELOOP` here.

I’ll be trying it out myself in a second, but I think you can safely keep all test changes from Trevor’s commit.

addaleax replied:
@bzoz Sorry, you’re right, that second `test_deep_symlink_eloop` part does not work with the JS implementation as-is. I think it can be fixed up to be included here, but that’s probably something better kept for another PR.

addaleax commented:
@bzoz Many, many thanks for starting to add regression tests! Overall LGTM, too.

bzoz commented:
CI run: https://ci.nodejs.org/job/node-test-pull-request/3453/

addaleax commented:
LGTM, thanks for doing the work behind this PR!

CI: https://ci.nodejs.org/job/node-test-commit/4374/

addaleax commented:
One Windows build bot failure, running CI again to be safe: https://ci.nodejs.org/job/node-test-commit/4376/

bzoz commented:
I had one running here https://ci.nodejs.org/job/node-test-commit/4373/, it is green.

addaleax commented:
Oh, sorry (but good to hear)! I’d say this is good to go once #7846 is resolved.

misterdjules commented:
@bzoz What do you think of squashing all commits but the ones from #7846 to make reviewing changes easier? Currently one has to go through all these commits separately in order to focus on reviewing the changes that are actually related to this PR, which is a bit tedious.

bzoz commented:
@misterdjules OK, squashed

Commits:

1. Restore old js realpath implementation
2. Implement new API (i.e. options.encoding) and add a test
3. Add tests and benchmarks from #7548 
4. Remove one test from #7548 which was failing

saghul commented:
What was the result of the discussion about the substed realpath behavior thing? This PR adds a test thus making certain behavior "correct" which might not necessarily be the case. It depends on how substed drives are interpreted, IMHO.

On the maze of issues and notes I recall reading that some people where using substed drives to get shorter paths. Was that the case? What breaks if we would give them the long path?

bzoz commented:
This is the way it worked with old implementation, and is one of the things that new C implementation broke. People rely on this to work this way, so if we ever change how realpath works again - like restoring C implementation in v7 - we will know that it is breaking change.

As for substed drives - it helps with "filename too long" errors, which happen on Windows.

bzoz commented:
CI after squashing is still green: https://ci.nodejs.org/job/node-test-pull-request/3518/

misterdjules commented:
This PR adds one new test, which is great. Does it make sense to add more tests, especially regression tests for https://github.com/nodejs/node/issues/7175, https://github.com/nodejs/node/issues/6978 and https://github.com/nodejs/node/issues/6624?

misterdjules commented:
Is there a reason why the last 4 commits of this PR are still not squashed into a minimum number of commits? In this case it seems it could be squashed into two commits : one that fixes the problem and the other that adds the benchmark. Otherwise it's still a bit tedious to review the changes.

addaleax commented:
@misterdjules The non-fixup commits contain only code that has already been signed off on in some form, I think that’s an acceptable reason to split the commits while the PR is still open. f2384fe is just a partial revert and reinstantiates old code, and d591e48 is taken from #7548.

saghul commented:
> This is the way it worked with old implementation, and is one of the things that new C implementation broke. People rely on this to work this way, so if we ever change how realpath works again - like restoring C implementation in v7 - we will know that it is breaking change.
> 
> As for substed drives - it helps with "filename too long" errors, which happen on Windows.

I disagree. The fact that it worked that way doesn't make it correct. At any rate, I brought this up in yesterday's CTC meeting and it was agreed upon that the test will make it in and we'll reevaluate if need be in the future.

addaleax commented:
@bzoz You can drop Myles’ reverts here now, they have landed and at least for me locally rebasing this PR against master just removes the commits.

bzoz commented:
Rebased, CI is here: https://ci.nodejs.org/job/node-test-pull-request/3534/

jasnell commented:
Before we land this we should invite folks to do some additional testing.
CITGM: https://ci.nodejs.org/view/Node.js-citgm/job/thealphanerd-smoker-xunit/42/

addaleax commented:
@jasnell What kind of testing did you have in mind?

jasnell commented:
@addaleax ... just making doubly sure that modules such as `glob` that depend on the realpath impl aren't broken by this.

bzoz commented:
Some CITGM tests failed, but those are unrelated. So - OK to land this?

jasnell commented:
Just noticed one unintended change on this... the ability to pass the path *in* as a `Buffer` is reverted in this also. For instance,

in Node v6.3.1:
```js
$ node
> fs.realpath(Buffer.from('/'))
undefined
> fs.realpathSync(Buffer.from('/'))
'/'
>
```

With this patch:
```js
> fs.realpath(Buffer.from('/'))
TypeError: Path must be a string. Received <Buffer 2f>
    at assertPath (path.js:7:11)
    at Object.resolve (path.js:1148:7)
    at Object.realpath (fs.js:1696:18)
    at repl:1:4
    at sigintHandlersWrap (vm.js:22:35)
    at sigintHandlersWrap (vm.js:96:12)
    at ContextifyScript.Script.runInThisContext (vm.js:21:12)
    at REPLServer.defaultEval (repl.js:307:29)
    at bound (domain.js:280:14)
    at REPLServer.runBound [as eval] (domain.js:293:12)
> fs.realpathSync(Buffer.from('/'))
TypeError: Path must be a string. Received <Buffer 2f>
    at assertPath (path.js:7:11)
    at Object.resolve (path.js:1148:7)
    at Object.realpathSync (fs.js:1597:18)
    at repl:1:4
    at sigintHandlersWrap (vm.js:22:35)
    at sigintHandlersWrap (vm.js:96:12)
    at ContextifyScript.Script.runInThisContext (vm.js:21:12)
    at REPLServer.defaultEval (repl.js:307:29)
    at bound (domain.js:280:14)
    at REPLServer.runBound [as eval] (domain.js:293:12)
>
```

This PR needs to be updated to preserve the ability to pass in the path as a `Buffer`

bzoz commented:
@jasnell good catch. I've updated the code to handle buffer parameters.

I've also added test for buffer as parameter. It also test all `encoding` options. CI is green: https://ci.nodejs.org/job/node-test-commit/4470/

jasnell commented:
@bzoz ... thank you. There is still a concern here in that calling `.toString('utf8')` means that the `Buffer` passed in must already be utf8 and the reason we now allow path as a `Buffer` is because it's not always the case that a path *only* contains utf8. Unfortunately, given that we're returning to the pure js implementation of realpath, there's not a really clear way of handling this without making much more significant internal changes, which we shouldn't do at this point. Hmm.. we'll just have to make sure the limitation is documented here when we communicate about this change...

bzoz commented:
I've added a note to the `fs` documentation.

jasnell commented:
LGTM, thank you.

jasnell commented:
CI is green. Let's give it another day before landing. @nodejs/ctc @nodejs/collaborators ... please take one final look.

saghul commented on lib/fs.js:
```diff
@@ -1563,38 +1563,239 @@ fs.unwatchFile = function(filename, listener) {
 
 
-fs.realpathSync = function realpathSync(path, options) {
+// Regexp that finds the next partion of a (partial) path
```
typo: partion

bzoz replied:
fixed

bzoz commented:
Landed in https://github.com/nodejs/node/commit/08996fde3c08c0cfabc7cc3fe332d83f674ab099 and https://github.com/nodejs/node/commit/b9832eb3fe61fcbfb6031e3dfce5eff8567a479c

bpasero commented:
@bzoz @addaleax one thing I really liked from the behaviour of native `fs.realpath` is that it would return the actual casing of a path as it exists on disk. E.g. if there is an actual path `/foo/bar/some.txt`, running `fs.realpath("/foo/bar/some.txt")` would return `/foo/bar/some.txt`. Are there any plans to support this behaviour after going back to the old JS implementation?

addaleax commented:
@bpasero I guess you can open a separate issue for this? It sounds like something that would be okay to have.

bpasero commented:
@addaleax makes sense, opened https://github.com/nodejs/node/issues/8715
